src/python/pants/base/exception_sink.py

```diff
@@ -59,20 +217,73 @@ def _format_exception_message(cls, msg, pid):
       timestamp=cls._iso_timestamp_for_now(),
       args=sys.argv,
       pid=pid,
-      message=msg,
-    )
+      message=msg)
+
+  _traceback_omitted_default_text = '(backtrace omitted)'
 
   @classmethod
-  def log_exception(cls, msg):
+  def _format_traceback(cls, tb, should_print_backtrace):
+    if should_print_backtrace:
+      traceback_string = ''.join(traceback.format_tb(tb))
+    else:
+      traceback_string = cls._traceback_omitted_default_text
+    return traceback_string
+
+  _UNHANDLED_EXCEPTION_LOG_FORMAT = """\
+Exception caught: ({exception_type})
+{backtrace}
+Exception message: {exception_message}{maybe_newline}
+"""
+
+  @classmethod
+  def _format_unhandled_exception_log(cls, exc, tb, add_newline, should_print_backtrace):
+    exception_message = str(exc) if exc else '(no message)'
+    maybe_newline = '\n' if add_newline else ''
+    return cls._UNHANDLED_EXCEPTION_LOG_FORMAT.format(
+      exception_type=type(exc),
+      backtrace=cls._format_traceback(tb, should_print_backtrace=should_print_backtrace),
+      exception_message=exception_message,
+      maybe_newline=maybe_newline)
+
+  @classmethod
+  def _exit_with_failure(cls, terminal_msg):
+    # Exit with failure, printing a message to the terminal (or whatever the interactive stream is).
+    cls._exiter.exit(result=cls.UNHANDLED_EXCEPTION_EXIT_CODE,
+                     msg=terminal_msg,
+                     out=cls._interactive_output_stream)
+
+  @classmethod
+  def _log_unhandled_exception_and_exit(cls, exc_class=None, exc=None, tb=None, add_newline=False):
```

Comment:
This method is called from outside of the class, and should maybe be public.

It's also worth considering whether this should (mostly) stay in the Exiter, and just call back to the ExceptionSink for the purposes of writing to the exception logs...? I'm not very clear on that though, given the fact that they both call one another at various times.

Reply:
I think I like that idea a lot. I think it specifically should be done, along with the comment on the Exiter holding onto its output stream, in a followup. I will continue addressing your comments and seeing what can be done here without lots more code changes.

Reply:
Broke out #6596!